Feat data management (#41)
alexandra-thompson-dc/rff-data-gov-guidance: docs/data-management/storage-options.qmd

Reviewer: Line 43: Suggest rephrasing: An RFF account is required for access to the L:/ drive. Therefore, when working with external collaborators it may be necessary to store and share data in other ways (in addition to storing data on the L:/ drive).

Reviewer: Line 47: This paragraph is a little confusing. What is it mean "If access to data is the main consideration"? I think this should instead say something like:

For projects with external collaborators, we recommend using a combination of OneDrive and Github. Github can be used for sharing code and sufficiently small data (well below 100 MB), and OneDrive can be used from sharing raw and processed data files between RFF analysts working on RFF servers, and external collaborators working elsewhere.

 "We recommend" (to avoid passive tense)

Reviewer: Line 53: "This folder can be accessed from any computer logged into a OneDrive account." This is not literally true because you need permissions. I would suggest just dropping this sentence.

Line 54: OneDrive is accessible from the web on lab computers, right?

Reviewer: Line 61: I actually think the primary reason to mirror directory structures is to instruct other users where that file should be placed if they are copying the file to their own directory.

Reviewer: Line 67: I've generally avoided running code using data from a cloud storage folder, and it sounds to me like this could be a headache. I would probably be inclined to caution people against this, but if others have done this and think it has benefits, that's okay too.

Author: @jmwingenroth about Matt's comment "Line 67: I've generally avoided running code using data from a cloud storage folder, and it sounds to me like this could be a headache. I would probably be inclined to caution people against this, but if others have done this and think it has benefits, that's okay too."
What do you think?

Reviewer 2: I remember from our discussions that the main use case for storing data on the cloud was working with non-RFF folks. I don't point my code to Dropbox or Google Drive. If keeping the data in the repo isn't feasible, I'd use, in order of preference: i) an wrapper package in the language I'm using (e.g., dataRetrieval), ideally with a cache to avoid automatically re-downloading, ii) API queries (e.g. curl GET) within an `if` statement checking if the file exists to avoid re-downloading, iii) a manually downloaded version of the file (maybe copied from Dropbox).

Reviewer: Yeah that seems reasonable to me. I've only ever used options 2 and 3. I'd love to see an example for how option 1 would work. In all of these options, the data would be shared via Dropbox, but then _stored_ in a separate local directory where the repo is stored.

Reviewer 2: For options 1 and 2, I was talking about not using the project's cloud storage. The code can be set up to download data from an external source. For instance, `dataRetrieval` queries the USGS/EPA WQP API (<- hell of an acronym). If the data was created as part of the project, or isn't able to be downloaded via curl/API, I think option 3 is the only one I've used.

Author: @alexandra-thompson-dc 
revise
"This folder can be accessed from any computer logged into a OneDrive account."
to "This folder can be accessed from any computer logged into the OneDrive account with the proper permissions."

## Summary

* Store data and code for RFF projects in a project-specific L:/ drive folder, and use GitHub to share and version control code.
* If working with external collaborators:
  * Use OneDrive to share select data. Only store data in OneDrive that is necessary for collaboration.
  * Use GitHub to share and review code.
  * For small datasets, GitHub may be used for both data and code storage.
* For both L:/ drive and OneDrive project folder setup, contact [IThelp@rff.org](mailto:ithelp@rff.org) with specifications.

## L:/ drive

The primary place to store project data and code is the L:/ drive.
All projects should have an L:/ drive folder, even when working with external collaborators.
It is optimized for data-intensive work, has large storage capacity, is regularly backed up, and is more secure than personal drives.
An RFF account is required to access the network, which can be accessed from work computers or remotely.
Git repositories can be setup within L drive project folders, as described in the Organization section below.

> ### L:/ drive storage request

Author: Good call! Done. Fun website: [random name generator](https://www.behindthename.com/random/)

## Summary

> * Store data and code for RFF projects in a project-specific L:/ drive folder.

Reviewer: "and use Github to share and version control code." I'd suggest saying this in order to avoid implying that Github should only be used for projects with external collaborators.

| Archival date | December 2026 |
| Data agreement or sensitive data security considerations | Proprietary data will be in raw data folder and will need to be made read-only |

> ## Working with external collaborators

Author: tag for group discussion

An RFF account is required for access to the L:/ drive. Therefore, when working with external collaborators it may be necessary to store and share data in other ways (in addition to storing data on the L:/ drive).
In that case, there are a few options.

### GitHub and OneDrive
For projects with external collaborators, we recommend using a combination of GitHub and OneDrive. Github can be used for sharing code and sufficiently small data (well below 100 MB), and OneDrive can be used from sharing raw and processed data files between RFF analysts working on RFF servers, and external collaborators working elsewhere.

#### Sharing code

> GitHub's distributed version control system enables users to collaboratively develop, organize, and share code across multiple computers. Unlike personal cloud storage services such as OneDrive, sharing code via GitHub allows users to maintain independent local versions of a project's codebase, using version control to track edits, reconcile differences, and seamlessly synchronize changes across devices. On the other hand, a disadvantage of using GitHub for sharing is that it does not support sharing files larger than 100MB; files above the size limit need to be shared in other ways.

Reviewer 2: ```suggestion
GitHub's distributed version control system enables users to collaboratively develop, organize, and share code across multiple computers. Unlike personal cloud storage services such as OneDrive, sharing code via GitHub allows users to maintain independent local versions of a project's codebase, using version control to track edits, reconcile differences, and seamlessly synchronize changes across devices. GitHub is intended for code and small data; files larger than 100 MB should be shared using other tools.
```

> 

Reviewer: This subsection is called "Collaborative coding", which I think is a little confusing; my first thought when reading it was that it referred to coding in teams. I'd suggest calling this "Sharing code", and the calling the other section "Sharing data"

Reviewer: Also, I would suggest putting this section above the OneDrive section. I think the way we suggest using OneDrive becomes a little clearer if you first know that most sharing is happening through Github.

> Use GitHub as the primary storage solution.

Reviewer: "primary tool for sharing code and, where possible, data."? This is a little confusing because we've earlier told people L:/ drive is the primary storage solution.

> GitHub repositories can be "cloned" to a local computer.

Reviewer: Here's an alternative version of this paragraph:

GitHub's distributed version control system enables users to collaboratively develop, organize, and share code across multiple computers. Unlike personal cloud storage services such as OneDrive, GitHub allows users to maintain independent local versions of a repository, using version control to track edits, reconcile differences, and seamlessly synchronize changes across devices. On the other hand, a disadvantage of using GitHub for sharing is that it does not support sharing large files. Files larger than 10GB need to be shared in other ways.

This method leverages GitHub's distributed version control features to streamline collaboration, organization, and resilience by allowing users to "clone" GitHub repositories to a local computer (such as an RFF server folder or personal OneDrive folder), make changes independently, and then upload those changes to a shared workspace that can be accessed through any web browser or any computer with git installed.

> As with data organization, use the same folder structure for code across shared folders.

Author: Matt, has this been resolved with the latest updates?

> As with data organization, use the same folder structure for code across shared folders.

Reviewer: No, it wasn't really. I just pushed an update to this section that deleted the first and last paragraphs of this section, which I think we're confusing and unnecessary. Let me know if you think I'm getting that wrong.

* __Security__: Consider the security implications of storing sensitive data on the cloud and specify access accordingly.

### Other options for collaborating with external partners

#### Accessing raw data via Application Program Interfaces (APIs)

APIs enable users or programs to download relevant subsets of data directly into memory, eliminating the need to download and store entire datasets on disk. This approach not only reduces storage requirements but also improves efficiency by allowing applications to process relevant data immediately without managing large raw files.

For large datasets, it is best to include lines in your code for downloading raw data from the web via an API, where available, rather than downloading data to disk and importing it.

> If an API is not available, having code download data from remote servers via other methods (e.g., `curl`) is still preferable.

Reviewer 2: ```suggestion
If an API is not available, having code download data from remote servers via other methods (e.g., `curl`) can also be a good option.
```

Using APIs is ideal for a variety of reasons:

* __Storage size__: An API allows the user/program to access specific data of relevance without needing to find room to store an entire, broader dataset (which is instead hosted on the data provider's server)
* __Reproducibility__: Using public APIs or other ways of accessing public data on the internet allows for code produced using other best practices to run "out of the box" on any computer/VM with internet access

> * __Updates__: Code that uses API queries retrieves the most current data available each time it is executed, removing the need to manually update a local dataset.

Reviewer 2: ```suggestion
* __Updates__: Code that uses API queries can be programmed to retrieve the most current data available each time it is executed, removing the need to manually update a local dataset.
```

#### Accessing raw data via Application Program Interfaces (APIs)

> For large datasets, it is best to include lines in your code for downloading raw data from the web via an API, where available, rather than downloading data to disk and importing it.

Reviewer 2: ```suggestion
Writing code that uses an API is especially advantageous for working with large datasets, because it often allows for subsets of the data to be loaded (e.g., part of a state map covering one city), which can reduce the need for local disk space.
```
